File: python-package/src/pijush_calculator/calculator.py
"""Advanced, dependency-free numerical engine for pijush-calculator."""
import math
import statistics

def _finite(value):
    if isinstance(value, float) and not math.isfinite(value):
        raise ValueError("result must be finite")
    return value

def add(a,b): return _finite(a+b)
def subtract(a,b): return _finite(a-b)
def multiply(a,b): return _finite(a*b)
def divide(a,b):
    if b==0: raise ZeroDivisionError("cannot divide by zero")
    return _finite(a/b)
def modulo(a,b):
    if b==0: raise ZeroDivisionError("cannot modulo by zero")
    return _finite(a%b)
def power(a,b): return _finite(a**b)
def percentage(value,percent): return _finite(value*percent/100)
def absolute(value): return _finite(abs(value))
def minimum(*values):
    if not values: raise ValueError("minimum requires at least one value")
    return min(values)
def maximum(*values):
    if not values: raise ValueError("maximum requires at least one value")
    return max(values)
def average(*values):
    if not values: raise ValueError("average requires at least one value")
    return _finite(statistics.fmean(values))
def mean(*values): return average(*values)
def sum_values(*values): return _finite(math.fsum(values))
def product(*values): return _finite(math.prod(values))
def clamp(value,minimum_value,maximum_value):
    if minimum_value>maximum_value: raise ValueError("minimum_value cannot exceed maximum_value")
    return min(maximum_value,max(minimum_value,value))
def reciprocal(value):
    if value==0: raise ZeroDivisionError("cannot take reciprocal of zero")
    return _finite(1/value)
def square(value): return _finite(value*value)
def cube(value): return _finite(value*value*value)
def square_root(value):
    if value<0: raise ValueError("square root requires a non-negative value")
    return _finite(math.sqrt(value))
def cube_root(value): return _finite(math.cbrt(value) if hasattr(math,"cbrt") else math.copysign(abs(value)**(1/3),value))
def nth_root(value,n):
    if n==0: raise ValueError("root degree cannot be zero")
    if value<0 and int(n)%2==0: raise ValueError("even root requires a non-negative value")
    result=math.copysign(abs(value)**(1/float(n)),value) if value<0 else value**(1/float(n))
    return _finite(result)
def factorial(value):
    if not isinstance(value,int) or value<0: raise ValueError("factorial requires a non-negative integer")
    return math.factorial(value)
def gcd(a,b): return math.gcd(int(a),int(b))
def lcm(a,b): return math.lcm(int(a),int(b))
def sine(value,degrees=False): return _finite(math.sin(math.radians(value) if degrees else value))
def cosine(value,degrees=False): return _finite(math.cos(math.radians(value) if degrees else value))
def tangent(value,degrees=False): return _finite(math.tan(math.radians(value) if degrees else value))
def secant(value,degrees=False): return _finite(1/cosine(value,degrees))
def cosecant(value,degrees=False): return _finite(1/sine(value,degrees))
def cotangent(value,degrees=False): return _finite(1/tangent(value,degrees))
def arcsine(value,degrees=False): return _finite(math.degrees(math.asin(value)) if degrees else math.asin(value))
def arccosine(value,degrees=False): return _finite(math.degrees(math.acos(value)) if degrees else math.acos(value))
def arctangent(value,degrees=False): return _finite(math.degrees(math.atan(value)) if degrees else math.atan(value))
def atan2(y,x,degrees=False): return _finite(math.degrees(math.atan2(y,x)) if degrees else math.atan2(y,x))
def hyperbolic_sine(value): return _finite(math.sinh(value))
def hyperbolic_cosine(value): return _finite(math.cosh(value))
def hyperbolic_tangent(value): return _finite(math.tanh(value))
def logarithm(value,base=10):
    if value<=0 or base<=0 or base==1: raise ValueError("invalid logarithm domain")
    return _finite(math.log(value,base))
def log2(value):
    if value<=0: raise ValueError("log2 requires a positive value")
    return _finite(math.log2(value))
def log1p(value):
    if value<=-1: raise ValueError("log1p requires a value greater than -1")
    return _finite(math.log1p(value))
def expm1(value): return _finite(math.expm1(value))
def natural_log(value):
    if value<=0: raise ValueError("natural logarithm requires a positive value")
    return _finite(math.log(value))
def exponential(value): return _finite(math.exp(value))
def hypot(*values):
    if not values: raise ValueError("hypot requires at least one value")
    return _finite(math.hypot(*values))
def combinations(n,r):
    n,r=int(n),int(r)
    if n<0 or r<0 or r>n: raise ValueError("invalid combination range")
    return math.comb(n,r)
def permutations(n,r):
    n,r=int(n),int(r)
    if n<0 or r<0 or r>n: raise ValueError("invalid permutation range")
    return math.perm(n,r)
def median(*values):
    if not values: raise ValueError("median requires at least one value")
    return statistics.median(values)
def variance(*values):
    if not values: raise ValueError("variance requires at least one value")
    return _finite(statistics.pvariance(values))
def standard_deviation(*values): return _finite(math.sqrt(variance(*values)))
def range_values(*values):
    if not values: raise ValueError("range requires at least one value")
    return maximum(*values)-minimum(*values)
def is_close(a,b,rel_tol=1e-9,abs_tol=0.0): return math.isclose(a,b,rel_tol=rel_tol,abs_tol=abs_tol)
def approximately_equal(a,b,tolerance=1e-9): return abs(a-b)<=tolerance

File: java-package/src/main/java/io/github/tejasmk2/calculator/Calculator.java
package io.github.tejasmk2.calculator;

import java.math.BigDecimal;
import java.math.MathContext;
import java.util.Objects;

/** Advanced decimal and scientific arithmetic engine for Java applications. */
public final class Calculator {
    private static final MathContext MC = MathContext.DECIMAL128;
    private Calculator() {}
    public static BigDecimal add(BigDecimal a,BigDecimal b){return require(a).add(require(b),MC);}
    public static BigDecimal subtract(BigDecimal a,BigDecimal b){return require(a).subtract(require(b),MC);}
    public static BigDecimal multiply(BigDecimal a,BigDecimal b){return require(a).multiply(require(b),MC);}
    public static BigDecimal divide(BigDecimal a,BigDecimal b){require(a);require(b);if(b.signum()==0)throw new ArithmeticException("cannot divide by zero");return a.divide(b,MC);}
    public static BigDecimal modulo(BigDecimal a,BigDecimal b){require(a);require(b);if(b.signum()==0)throw new ArithmeticException("cannot modulo by zero");return a.remainder(b,MC);}
    public static BigDecimal power(BigDecimal a,int exponent){return require(a).pow(exponent,MC);}
    public static BigDecimal percentage(BigDecimal value,BigDecimal percent){return require(value).multiply(require(percent),MC).divide(BigDecimal.valueOf(100),MC);}
    public static BigDecimal absolute(BigDecimal value){return require(value).abs(MC);}
    public static BigDecimal minimum(BigDecimal... values){return minMax(values,true);}
    public static BigDecimal maximum(BigDecimal... values){return minMax(values,false);}
    public static BigDecimal average(BigDecimal... values){if(values.length==0)throw new IllegalArgumentException("average requires a value");return sum(values).divide(BigDecimal.valueOf(values.length),MC);}
    public static BigDecimal mean(BigDecimal... values){return average(values);}
    public static BigDecimal sum(BigDecimal... values){BigDecimal total=BigDecimal.ZERO;for(BigDecimal v:values)total=add(total,v);return total;}
    public static BigDecimal product(BigDecimal... values){BigDecimal total=BigDecimal.ONE;for(BigDecimal v:values)total=multiply(total,v);return total;}
    public static BigDecimal clamp(BigDecimal value,BigDecimal minimum,BigDecimal maximum){require(value);require(minimum);require(maximum);if(minimum.compareTo(maximum)>0)throw new IllegalArgumentException("minimum cannot exceed maximum");return value.max(minimum).min(maximum);}
    public static BigDecimal reciprocal(BigDecimal value){require(value);if(value.signum()==0)throw new ArithmeticException("cannot take reciprocal of zero");return BigDecimal.ONE.divide(value,MC);}
    public static BigDecimal square(BigDecimal value){return multiply(value,value);}
    public static BigDecimal cube(BigDecimal value){return multiply(square(value),value);}
    public static double squareRoot(double value){if(value<0)throw new IllegalArgumentException("square root requires a non-negative value");return Math.sqrt(value);}
    public static double cubeRoot(double value){return Math.cbrt(value);}
    public static double sine(double value,boolean degrees){return Math.sin(degrees?Math.toRadians(value):value);}
    public static double cosine(double value,boolean degrees){return Math.cos(degrees?Math.toRadians(value):value);}
    public static double tangent(double value,boolean degrees){return Math.tan(degrees?Math.toRadians(value):value);}
    public static double secant(double value,boolean degrees){return 1.0/cosine(value,degrees);}
    public static double cosecant(double value,boolean degrees){return 1.0/sine(value,degrees);}
    public static double cotangent(double value,boolean degrees){return 1.0/tangent(value,degrees);}
    public static double arcsine(double value,boolean degrees){double r=Math.asin(value);return degrees?Math.toDegrees(r):r;}
    public static double arccosine(double value,boolean degrees){double r=Math.acos(value);return degrees?Math.toDegrees(r):r;}
    public static double arctangent(double value,boolean degrees){double r=Math.atan(value);return degrees?Math.toDegrees(r):r;}
    public static double atan2(double y,double x,boolean degrees){double r=Math.atan2(y,x);return degrees?Math.toDegrees(r):r;}
    public static double hyperbolicSine(double value){return Math.sinh(value);}
    public static double hyperbolicCosine(double value){return Math.cosh(value);}
    public static double hyperbolicTangent(double value){return Math.tanh(value);}
    public static double logarithm(double value,double base){if(value<=0||base<=0||base==1)throw new IllegalArgumentException("invalid logarithm domain");return Math.log(value)/Math.log(base);}
    public static double naturalLog(double value){if(value<=0)throw new IllegalArgumentException("natural logarithm requires a positive value");return Math.log(value);}
    public static double exponential(double value){return Math.exp(value);}
    public static double hypot(double... values){if(values.length==0)throw new IllegalArgumentException("hypot requires a value");double sum=0;for(double v:values)sum+=v*v;return Math.sqrt(sum);}
    public static long gcd(long a,long b){return java.math.BigInteger.valueOf(a).gcd(java.math.BigInteger.valueOf(b)).longValue();}
    public static long lcm(long a,long b){if(a==0||b==0)return 0;return Math.abs(a/gcd(a,b)*b);}
    public static BigDecimal factorial(int n){if(n<0)throw new IllegalArgumentException("factorial requires a non-negative integer");BigDecimal result=BigDecimal.ONE;for(int i=2;i<=n;i++)result=result.multiply(BigDecimal.valueOf(i),MC);return result;}
    public static BigDecimal combinations(int n,int r){if(n<0||r<0||r>n)throw new IllegalArgumentException("invalid combination range");r=Math.min(r,n-r);BigDecimal result=BigDecimal.ONE;for(int i=1;i<=r;i++)result=result.multiply(BigDecimal.valueOf(n-r+i),MC).divide(BigDecimal.valueOf(i),MC);return result;}
    public static BigDecimal permutations(int n,int r){if(n<0||r<0||r>n)throw new IllegalArgumentException("invalid permutation range");BigDecimal result=BigDecimal.ONE;for(int i=0;i<r;i++)result=result.multiply(BigDecimal.valueOf(n-i),MC);return result;}
    public static BigDecimal median(BigDecimal... values){if(values.length==0)throw new IllegalArgumentException("median requires a value");BigDecimal[] copy=values.clone();java.util.Arrays.sort(copy);int mid=copy.length/2;return copy.length%2==1?copy[mid]:copy[mid-1].add(copy[mid],MC).divide(BigDecimal.valueOf(2),MC);}
    public static BigDecimal variance(BigDecimal... values){if(values.length==0)throw new IllegalArgumentException("variance requires a value");BigDecimal mean=average(values);BigDecimal total=BigDecimal.ZERO;for(BigDecimal v:values){BigDecimal d=v.subtract(mean,MC);total=total.add(d.multiply(d,MC),MC);}return total.divide(BigDecimal.valueOf(values.length),MC);}
    public static double standardDeviation(BigDecimal... values){return Math.sqrt(variance(values).doubleValue());}
    public static BigDecimal range(BigDecimal... values){return maximum(values).subtract(minimum(values),MC);}
    private static BigDecimal minMax(BigDecimal[] values,boolean min){if(values.length==0)throw new IllegalArgumentException("requires a value");BigDecimal result=require(values[0]);for(int i=1;i<values.length;i++){BigDecimal v=require(values[i]);result=min?result.min(v):result.max(v);}return result;}
    private static BigDecimal require(BigDecimal value){return Objects.requireNonNull(value,"value must not be null");}
}
